docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
+* all values are propagated to a role through `r_` prefixed variables only
+* all values that are set inside a role are read outside of a role through `r_` prefixed variables only
+* all available `r_` prefixed variables are documented and checked before a main role task is run
+* a table of `r_` prefixed variables in README.md is automatically generated
+
+### Terminilogy
+
+**Role variables**: All `r_` prefixed variables read/set inside a role
+
+**Local variables**: All `l_` prefixed variables set inside a role
+
+### Role properties
+
+**Solid API**: all variables that a role caller can set need to be well defined.
+It means each such variable must have the following information attached:
+* name
+* description
+* default value (if the variable is defined)
+* set of acceptable values
+* if it is required or if it is optional
+
+Well defined API allows automatically generated documentation and variable checks.
+
+**Free of outer variables**: only variables consisting the role API or defined locally are allowed.
```

Comment:
I agree that the we want to take this in from the inventory.  I'd prefer to not pass in a default since it becomes an artifact the moment it gets merged.  I'd prefer we pass it in and if its not set then explode.  Being explicit is better than implicit for a couple of reasons.  
- If I don't know about a variable then I'm guaranteed to get something installed but it might not be what I desire.  This is sometimes worse than doing anything at all.  I have installed cluster and received example.com as my domain in my config files.  This becomes a debugging nightmare and a reinstall is the only solution.
- fail fast (fail: when a var isn't set) is a great pattern for the roles.  Its a form of defensive programming.  Role variable checking is a good way to ensure that something doesn't get half way configured and then break.  Especially in production clusters.